### ISSUE TITLE
remove unimplemented nn_req_send / nn_req_recv

### ISSUE THREAD

**Issue by ano4r** [nanomsg/nanomsg#596]
Implementations of exported function nn_req_send and nn_req_recv are missing:

```
int nn_req_send (NN_UNUSED int s, NN_UNUSED nn_req_handle hndl,
    NN_UNUSED const void *buf, NN_UNUSED size_t len, NN_UNUSED int flags)
{
    nn_assert (0);
}

int nn_req_recv (NN_UNUSED int s, NN_UNUSED nn_req_handle *hndl,
    NN_UNUSED void *buf, NN_UNUSED size_t len, NN_UNUSED int flags)
{
    nn_assert (0);
}
```

Do you plan to support these functions later?
Is there any other way to send response to particular client?

**Comment by dangerousHobo** [comment]
I'm also curious about this. As currently Req/Rep is not working for me when one end is in C++ and the other in Go. When digging into the issue I found the exported functions. Hoping some work is be done here that may resolve my issue.

**Comment by wirebirdlabs** [comment]
I would love to see this ability fleshed-out. See: http://www.freelists.org/post/nanomsg/Asynchronous-raw-sockets-in-nanomsg

My understanding here for this feature is adding the ability to both REQ *and* REP socket topologies where they maintain a table (trie? hash table?) of their outstanding requests. This would supersede the current behavior of only allowing one at a time per socket, and effectively cancelling in-flight requests if a new request is generated before the response arrives.

Such behavior is only currently possible by using RAW sockets, and doing the bookkeeping at the application layer. This new feature would bake-in the ability to the library.

**Comment by wirebirdlabs** [comment]
> As currently Req/Rep is not working for me when one end is in C++ and the other in Go

Would you open a new issue? This is likely unrelated, but an important thing to track. thanks!

**Comment by gdamore** [comment]
When using raw mode sockets, there shouldn’t be any problems — the main
thing is that you have to figure how to manage multiple requests and any
cancelation (if any) in your application.  I don’t think there’s a single
right way to do this — its necessarily pretty application specification.

Request IDs are generated by the framework even for raw mode requests
(IIRC), so its just a matter of matching requests to replies.

Such matching really is tied to the application behavior — for example
should a response back to a request done asynchronously cause a callback to
fire, or should threads be used, or some kind of sub-select/poll be
possible on outstanding requests?

The answer also will vary largely based the runtime.  For example, runtimes
with threads may find that running multiple requests and then blocking on a
reply (perhaps using a single thread to poll for responses that then finds
the request and signals completion) quite natural.  In other runtimes
threads are simply not an option.

So, I’d discourage attempts to provide a silver bullet solution here.
Instead, a body of examples showing how to use the raw mode, perhaps with
samples that show asynchronous handling, would be better.  I know that I’ve
implemented this asynchronous handling for RPC servers written against both
nanomsg and mangos several times before.

(Real world concerns to think about — should there be a limit on the number
of outstanding requests?  Do requests complete synchronously on the server,
or can they take a long time?  Etc.)

**Comment by gdamore** [comment]
Coming back to this, I'd like to see some kind of API proposal -- I think what you really want here is more like a real RPC framework (or REQ/REP framework) where you wind up either registering a callback, or handling a bunch of messages in parallel (perhaps using threading in the server).  This would probably be pretty easy in the server side to do.  But figuring out how to do the REQ (client) mode asynchronously cleanly is *vastly* harder.  You need to have a way to figure out how to match the requests and responses, which is a pain.  A client library could probably give you back the request ID which could be used as a cookie/ticket to use when doing a modified recv.

But as I think about this, this really looks like it should be not in libnanomsg core, but perhaps as a secondary library on top of nanomsg.  In any event, its not something we're going to do for 1.0 -- perhaps post 1.0 we can explore ideas though.

**Comment by gdamore** [comment]
I'm going to remove these APIs for now.  We can reassess in the future, with a more properly baked API proposal - or with an additional library layer on top.

**Comment by gdamore** [comment]
As part of this, we are also removing the unused nn_req_handle structure.
